File: Java/middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode currentNode = head;
        
        ListNode slowPtr = head;
        ListNode fastPtr = head;
        
        while(fastPtr != null && fastPtr.next != null){
            slowPtr = slowPtr.next; // slow pointer will move 1 node at a time
            fastPtr = fastPtr.next.next; // fast pointer will move 2 nodes at a time
        }
        
        // at end slow pointer will be pointing to middle node
        return slowPtr;
    }
}

File: Python/Middle_of_the_Linked_List.py
#Time Complexity: O(n)
#Space Complexity: O(1)

# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast =  head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow
            
